src/zcash/NoteEncryption.cpp

```diff
@@ -48,6 +101,155 @@ void KDF(unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE],
 
 namespace libzcash {
 
+boost::optional<SaplingNoteEncryption> SaplingNoteEncryption::FromDiversifier(diversifier_t d) {
+    uint256 epk;
+    uint256 esk;
+
+    // Pick random esk
+    librustzcash_sapling_generate_r(esk.begin());
+
+    // Compute epk given the diversifier
+    if (!librustzcash_sapling_ka_derivepublic(d.begin(), esk.begin(), epk.begin())) {
+        return boost::none;
+    }
+
+    return SaplingNoteEncryption(epk, esk);
+}
+
+boost::optional<SaplingEncCiphertext> SaplingNoteEncryption::encrypt_to_recipient(
+    const uint256 &pk_d,
+    const SaplingEncPlaintext &message
+)
+{
+    if (already_encrypted_enc) {
+        throw std::logic_error("already encrypted to the recipient using this key");
+    }
+
+    uint256 dhsecret;
+
+    if (!librustzcash_sapling_ka_agree(pk_d.begin(), esk.begin(), dhsecret.begin())) {
+        return boost::none;
+    }
+
+    // Construct the symmetric key
+    unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE];
+    KDF_Sapling(K, dhsecret, epk);
```

Comment:
This puts the output parameter first, whereas for ``librustzcash_sapling_ka_agree`` for example it is last. I don't know which one should change; ``crypto_aead_chacha20poly1305_ietf_encrypt`` also puts the output parameter first.